design/authenticators/application-identity/application-identity.md

```diff
@@ -0,0 +1,464 @@
+  3. Verify the Host Application Identity matches the authentication request
+   using the:
+     - Host Application Identity object built in the first step
+     - Authentication request Application Identity object given in the input
+      to this class
+      
+Let's see how the classes will look, in high level.
+
+#### Application Identity class
+
+We will have 2 different Application Identity classes. Although their
+interface is the same, they are not entirely identical. 
+
+At first, the `constraints` method will return a simple key-value hash. We
+should challenge that and find a better way to do it in Ruby.
+
+##### Host Application Identity
+
+```
+class HostApplicationIdentity
+
+      def initialize(authenticator_name:, service_id:, role_annotations:, permitted_annotations:)
+        @authenticator_name = authenticator_name
+        @service_id       = service_id
+        @role_annotations = role_annotations
+        @permitted_annotations = permitted_annotations
+      end
+
+      def constraints
+        @constraints ||= {}
+
+        @permitted_annotations.each { |annotation_name|
+          @constraints[annotation_name] = constraint_value(annotation_name)
+        }
+      end
+
+      private
+
+      # check the `service-id` specific constraint first to be more granular
+      def constraint_value constraint_name
+        annotation_value("#{@authenticator_name}/#{@service_id}/#{constraint_name}") ||
+          annotation_value("#{@authenticator_name}/#{constraint_name}")
+      end
+
+      def annotation_value name
+        annotation = @role_annotations.find { |a| a.values[:name] == name }
+
+        # return the value of the annotation if it exists, nil otherwise
+        if annotation
+          @logger.debug(LogMessages::Authentication::RetrievedAnnotationValue.new(name))
+          annotation[:value]
+        end
+      end
+    end
+```
+
+##### Authentication request Application Identity
+
+The input for this class will be built in each authenticator as it differs
+ between them. For simplicity, the class will receive the `constraints` hash
+  as an input. We should challenge that to find a better way.
+```
+class AuthenticationRequestApplicationIdentity
+  attr_reader :constraints
+
+  def initialize(constraints:)
+    @constraints = constraints
+  end
+end
+```
+
+#### ValidateApplicationIdentity class
+
+```
+ValidateApplicationIdentity ||= CommandClass.new(
+  dependencies: {
+    host_application_identity_class: HostApplicationIdentity
+  },
+  inputs:       %i(
+    authenticator_name
+    service_id
+    host_annotations
+    permitted_constraints
+    required_constraints
+    non_permitted_combinations
+    authentication_request_application_identity
+  )
+) do
+
+  def call
+    extract_application_identity_from_host
+    verify_configuration
+    verify_application_identity_matches_authentication_request
+  end
+
+  def extract_application_identity_from_host
```

Comment:
Correct me if I am wrong but I feel like `extract_application_identity_from_host` is not really representative of what this function is doing. `.new` just initializes the object but doesn't extract anything. Only host_application_identity.constraints "extracts" and checks if the annotations present are permitted

Maybe `initialize_application_identity_from_host`?

Reply:
i don't want to go into this implementation detail.

Reply:
but you are correct :)